File: src/algorithms/dynamic-programming/unique-paths/sources/unique-paths.py
# Unique Paths (Tabulation) — count distinct paths from top-left to bottom-right in a rows×columns grid
def unique_paths(rows: int, columns: int) -> int:  # @step:initialize
    dp_table = [1] * columns  # @step:initialize,fill-table
    # First row is all 1s — only one way to reach any cell by moving right only
    for row_index in range(1, rows):  # @step:compute-cell
        for column_index in range(1, columns):  # @step:compute-cell
            dp_table[column_index] = dp_table[column_index] + dp_table[column_index - 1]  # @step:compute-cell,read-cache
    return dp_table[columns - 1]  # @step:complete

File: src/algorithms/dynamic-programming/counting/unique-paths/sources/UniquePaths.java
// Unique Paths (Tabulation) — count distinct paths from top-left to bottom-right in a rows×columns grid
public class UniquePaths {
    public static int uniquePaths(int rows, int columns) { // @step:initialize
        int[] dpTable = new int[columns]; // @step:initialize,fill-table
        java.util.Arrays.fill(dpTable, 1); // @step:fill-table
        // First row is all 1s — only one way to reach any cell by moving right only
        for (int rowIndex = 1; rowIndex < rows; rowIndex++) { // @step:compute-cell
            for (int columnIndex = 1; columnIndex < columns; columnIndex++) { // @step:compute-cell
                dpTable[columnIndex] = dpTable[columnIndex] + dpTable[columnIndex - 1]; // @step:compute-cell,read-cache
            }
        }
        return dpTable[columns - 1]; // @step:complete
    }
}
